### ISSUE TITLE
Adding isfshax as an advanced/alternative exploit

### ISSUE THREAD

**Issue by jbmagination** [hacks-guide/Guide-WiiU#236]
Me and some others were talking about this so I figured it would be worth making a GitHub issue about since I fully expect we're not the only ones thinking about it.

I think [isfshax](https://github.com/isfshax/isfshax) should be added to the guide - but NOT as the main exploit. I have gripes with it that make it difficult for me to fully sell it as the main exploit for what is largely intended to be a noob-friendly guide.

* Autobooting cannot be setup within the minute_minute GUI - it must be done as a configuration file on your SD card. If you *don't* set it up, you must use your Power and Eject buttons to navigate which option you want - without access to the GamePad - every time you boot up your Wii U. You also cannot use autobooting without your SD card inserted - the minute_minute GUI will appear every time. 
* Enables sigpatches, which were on the guide before when Tiramisu was a thing but were removed when the guide started mainly recommending Aroma instead - and I think that's the right choice. With the Homebrew on Menu plugin, sigpatches are basically unnecessary (unless I'm forgetting something important!)
* Potential SLC corruption. It is considered quite safe, but for this reason, people who want to set it up are currently still advised to make a backup of their SEEPROM, OTP, and SLC.
* Hynix MLCs - who need something like this the most - only make up ~2.5% of Wii U's on the [Wii U System Database](https://wiiu.gerbilsoft.com/?sort=mlc_mid_prv&direction=desc) (link is sorted by MLC). For everyone else, I feel that it's largely not worth the hassle.

However, if they do set it up - especially if they have Hynix MLCs, like myself - and don't fuck things up in the process of setting it up, which I nearly did more than once - it serves as very good brick protection. I am one who likes to tinker and break and touch things I really shouldn't be touching, and isfshax with some backups has already saved my ass quite a few times. It also provides a minimal CFW even if you don't choose to setup direct PayloadLoader access into something like Aroma.

**Comment by VannyBuns** [comment]
We already have ISFShax on the wiki. There's just no place for it on the guide. Even so, it's up to @Lazr1026

**Comment by Lazr1026** [comment]
I do want isfshax to be on the main guide _eventually_, but I don't see it as absolutely necessary. It's main purpose atm is just to recover consoles

> Autobooting cannot be setup within the minute_minute GUI  

You can set up the fastboot minute on the SLC, which will init the bare minimum for minute to function _and_ automatically load plugins from the SLC and boot the console, but if something goes wrong with haxcopy, users will have to deal with ftp and users are not very smart most of the time..

> sigpatches are basically unnecessary   

The only use they have anymore that isn't straight piracy is for VC injects.

> Potential SLC corruption  

While this is a valid concern, the isfshax superblocks are installed to the very last superblocks on the SLC NAND, so I doubt there will be any corruption with it if the SLC isn't already dying.

> Hynix MLCs  

You mean you don't like playing with fire!?

**Comment by jbmagination** [comment]
> The only use they have anymore that isn't straight piracy is for VC injects.

AFAIK that's not on the guide or the wiki; correct me if I'm wrong. My main concerns are someone installing malicious homebrew that tampers with the system (which there are still avenues for, but two - the Homebrew on Menu and SDCafiine plugins - are better than three) and piracy

> You mean you don't like playing with fire!?

I once considered uninstalling OSv10 and replace it with an older one. I was about to actually do it, and then alarm bells went off in my head so I looked up what I was about to do before doing it and learned from [your friend's blog post](https://nightkingale.com/posts/the-downgrade-of-doom.html) that I was messing with the wrong application and that that might not be a good idea :fire: 

> but if something goes wrong with haxcopy, users will have to deal with ftp and users are not very smart most of the time..

See previous comment

**Comment by danny8376** [comment]
> AFAIK that's not on the guide or the wiki;
It is on uwuvci’s guide page, as like said above, this is the only legit case you need sigpatch now

**Comment by jan-hofmeier** [comment]
> You also cannot use autobooting without your SD card inserted - the minute_minute GUI will appear every time.

If you have minute on the SLC it will autoboot with the plugins from the SLC, if no SD is detected.
If neither the SD nor a fw.img on the SLC is detected by ISFShax it will boot with minimal patches. To most users that would look stock, the only difference is in the eco mode or IOSU reload behavior. Meaning even if standby services are enabled, the console will shutdown properly and turn off the dram. It will take a few seconds longer to boot, then starting from the standby mode, where iosu is already in dram. ECO mode does still work. 
When you leave the settings a stock system does an IOSU reload, with the minimal patches a full reboot is done, which takes a little longer

> Enables sigpatches, which were on the guide before when Tiramisu was a thing but were removed when the guide started mainly recommending Aroma instead - and I think that's the right choice. With the Homebrew on Menu plugin, sigpatches are basically unnecessary (unless I'm forgetting something important!)

I am not 100% sure why stroopwafel has them and if they can be removed without breaking anything. Apparently they aren't even complete since someone complained about them not working properly. I don't intend to fix them. 
One legit use case is to use them to launch an installed Homebrew Channel which gives you exec on the PCC without additional CFW. For ARM stuff, there could be a mocha stroopwafel plugin. That even already exists as a PoC. But this usecase probably isn't really relevant since we have Aroma

> Potential SLC corruption. It is considered quite safe, but for this reason, people who want to set it up are currently still advised to make a backup of their SEEPROM, OTP, and SLC.

The wiiu.hacks.guide also tells users to backup the same things. Having otp and seeprom is useful for recovering stuff from USB even if the console dies completely. SLC is also nice to have, but we never needed it. 
SLC corruption that we saw was usually a result of iosu having the wrong OTP and therefore wrong SLC keys. That happened for example with defuse when the otp on the SD was missing or running a stroopwafel without otp redirection on or when running a stroopwafel with otp redirection but minute not supplying the otp in the right way. To prevent that the mechanism to pass the otp was changed and newer minutes lock out older stroopwafels. The redirection now will only happen if mi Ute really supplies a otp in the right way. The newer minute usually doesn't supply a otp on ISFShax, except with an option for redNAND and only if also the SLC is redirected too.
In addition the latest isfshax plugin blocks the SCFM from formatting the SLC (which it tries to do when the wrong otp is supplied. This block is also included in the minimal patches that happen if no fw.img is found.

ISFShax always felt more risky to me too, but rationally I could argue it's safer than coldbooting Payloader. ISFShax requires just the writing of a working superblock. If the write failes, the hmac is wrong and it will just be ignored. Also a user can't easily supply a wrong superblock, since the installer checks the sha which is supplied in a second file and will refuse to install in case of a mismatch.

With the Payloadloader the system.xml has to be changed, which is arguably a more complex operation. If the system.xml is broken you will need defuse. (But that doesn't mean coldbooting Payloader is unsafe, Maschell put a lot of effort into checks and making sure everything is correct, that we can all take a moment to appreciate. Also there are no bricks caused by that known to me).

Also if people use homebrew to mess with system files over ftp, they better have ISFShax than not. There are enough bricks caused by people accidentally deleting something important or breaking the home menu or something like that. 


I already talked with Maschell about using ISFShax as the entry exploit and I think the as that would be the long-term goal. But to take full advantage of ISFShax we wouldn't just use it as the entry exploit but also use it to load mocha as a stroopwafel plugin to get rid of the IOSU exploit and to patch Cafe OS to get rid of the cafe is exploit. That then would require some changes that would likely deprecate running aroma from payloadloader or the the browser exploit (else we would have to maintain two mocha versions )

So maybe we should just wait a little more and see what @Maschell thinks. But I don't think this has a high priority to him atm since he is working on other cool stuff

**Comment by team-orangeBlue** [comment]
Hey guys it's been a year.

Any ideas?
ISFShax by this point is a must, a fair share of "quality-of-life" fixes mandate exploiting boot1 to introduce them, and the easiest gateway for them is isfshax.
Not to mention a fair share of bricks that could have been avoided if it was "mandatory" to install.

**Comment by zer00p** [comment]
I think this will settle the discussion, once it is out of beta: https://wafel.xyz/
https://github.com/zer00p/wafel_installer

**Comment by suprstarrd** [comment]
I was looking for something like this @zer00p but I'm extremely hesitant after learning it was co-created by AI. I agree; this needs a test / demonstration before I can safely recommend.

**Comment by team-orangeBlue** [comment]
Can someone with PTB/defuse run this and see if it kills the console or not please?

**Comment by zer00p** [comment]
I am glad you are bringing these concerns up, but they have me now concerned what will happen to the wii.hacks.guide? Basically everything in the guide is linked against libogc: https://github.com/fail0verflow/hbc/blob/master/README.md

Also I am taking your accusations of violating Nintendo’s copyright very seriously. So can you please point out the parts of the code where you suspect a copyright infringement could have happened? This code makes use of the low level functionality provided by wut and doesn't implement functionality usually provided by the SDK itself. Also there is no functionality in it, that you can reasonably expect to come from a Nintendo software, which wasn't already in the dumpling base.

Also what are your plans on the upcoming Aroma release, which will likely also contain code AI generated code (looking at some of the repos)? Will Aroma be removed from the guide or will you change the instructions to use some back level version of Aroma?

Will users have to rely on other sources like video guides on youtube?

**Comment by TechnoEggnog-2** [comment]
I had a brief discussion with some of the Wii U homebrew devs in a discord server.

A handful of them seem to agree on the idea that “AI Assisted” coding and work can safely be used in projects like these, mostly in writing simple boiler plate code. It’s one thing if people are using AI to generate their entire project, but it seems like a shift is being made to use it more appropriately.

With a pretty small amount of people who actually develop for the console… it would be hard to justify alternatives for those who truly will not accept any AI. Even just older versions of stuff will eventually miss out on improvements and features. I’d argue the guide will continue to use these projects.

**Comment by suprstarrd** [comment]
@TechnoEggnog-2 I think that's totally fair, but "co-created" to me is not the same as minor assistance. If that *is* how it was used, then that should be clarified! I have no qualms there as long as it's made clear on the installer/downloader page that some AI was used in limited quantity; the guide itself need not mention it.

**Comment by zer00p** [comment]
It's 90% vibe coded

**Comment by TechnoEggnog-2** [comment]
…and now I wonder if the AI stole code from dumpling directly after hearing this. I had thought you were using AI appropriately but now it’s obvious you just prompted it.

I can’t defend you from these people anymore unfortunately because I’d be advocating for a full AI takeover when that’s pretty explicitly not what I want AI to become.

**Comment by zer00p** [comment]
> …and now I wonder if the AI stole code from dumpling directly after hearing this

It is forked from dumpling, which should be obvious by the fact that the repo says "forked from dumpling-app/dumpling", the fact that the LICENSE mentions the orignal authors, the fact that the README credits
> [Crementif](https://github.com/Crementif) for the original [dumpling](https://github.com/dumpling-app/dumpling)
> [emiyl](https://github.com/emiyl) for [dumpling-classic](https://github.com/emiyl/dumpling-classic)

And the app printing this on startup: https://github.com/zer00p/wafel_installer/blob/5df858e2684357e02241242caafb4b8c9cc288fd/source/app/menu.cpp#L28

The MIT License of the original dumpling grants me and everyone else the rights to use it in this way. It doesn't even require attribution just retaining of the LICENSE, which I have done. https://github.com/dumpling-app/dumpling/blob/master/LICENSE.md

So I would appreciate if you could maybe do the tiniest bit of research before you accusing me of theft.

Also Crementif knows of this and even supported me, for which I am grateful.


Also you don't have to defend me from anyone. People can use it if they want. No one has to. If this guide want to stay legacy, that will be fine, people will find an up to date guide elsewhere

**Comment by ihaveamac** [comment]
If 90% of your project was generated by an LLM, how do we know you are even capable of maintaining it? Do you understand what it does? How do you know it did not take code from elsewhere without attribution?

**Comment by TechnoEggnog-2** [comment]
A loose transcription of what the “developer” did:

>forked dumpling
>opened google jules
>”rewrite this so that it installs isfshax instead”
>output
>proceeds to not test and ships it

**Comment by zer00p** [comment]
Oh, it is using code from elsewhere. I told it too, you would have found it in the LICENSE.md if you actually cared. The rest is too specific to this project and written to the spec of the prompt. There really isn't much else to it, that you could straight up take from another project. All the generic stuff is already provided by the wut librarys, dumpling and the included library.

And of course I understand what it does and it does exactly what I want it to do.


> A loose transcription of what the “developer” did:
> 
> > forked dumpling
> > opened google jules
> > ”rewrite this so that it installs isfshax instead”
> > output
> > proceeds to not test and ships it

This is naive at best and a malicious insinuation at worst

**Comment by TechnoEggnog-2** [comment]
> This is naive at best and a malicious insinuation at worst

What's really malicious is your intent to market this as the all-in-one solution when this is at-best a demo project. The risks involved with maintaining and distributing largely AI generated content are too high for us to handle.

Thanks for your work, but I don't believe this solution is the best fit.

**Comment by zer00p** [comment]
> What's really malicious is your intent to market this as the all-in-one solution

what about this is malicious? Did I make any wrong claim? does it not do, what I said it does? Did I lie about anything?
You are coming here with one accusation after another once the last one was addressed and nothing to back them up.

Also did you actually review the code, since you claim it isn't maintainable? I found it quite maintainable, but feel free to point out the parts that you don't like

**Comment by suprstarrd** [comment]
> The risks involved with maintaining and distributing largely AI generated content are too high for us to handle.

I am now seeing that in that case, isfshax may not be not acceptable for the guide? [Looking at the isfshax contributors chart](https://github.com/isfshax/isfshax_installer/graphs/contributors), the top 2 contributors are Jan and the Google Jules AI coding assistant (which [was also used to create wafel.xyz](https://github.com/zer00p/wafel_installer/graphs/contributors)). Google Jules has made 13 commits to isfshax. Jan only co-authored one of those commits. Jan, as a human contributor, still makes regular (and the majority of) commits.

This bleeds over into #280 - I don't think we can get away with vibing (no pun intended) our way to defining what "near-fully" AI generated code looks like.

**Comment by jan-hofmeier** [comment]
> Google Jules has made 13 commits to isfshax. Jan only co-authored one of those commits

Not that it would matter, but all commits there have me as co-auther. But that means absolutely nothing, it's just a setting in Jules.
Also going by the number of commits is just a bad metric. You can have huge commits and commits just fixing a typo, the number says nothing. rw has just one commit, but he contributed the majority of the code, with my contributions not even getting close (given much of that code came originally from minute or mini). 
If you work with something like jules it basically does one commit for every prompt. If you are not happy with something it did, because you actually reviewed the changes, you prompt it again (or comment the code in the review, which Jules can also pick up) which will then cause new commits. The closer you work with the code, the more commits you will get, if you don't squash them before merging (which makes counting the commits even more irrelevant). Also if I use geminicli locally, you would never see it in the commit message...

But I am not even sure if I want to associate with this guide any more (and I never asked in the past for it to be included).  I feel a certain hostility from a particular vocal group of people who want to tell you what tools you are allowed to use in your projects. Even tough these people don't seem to a clue how these tools are used, what they are capable off and what not.
Maybe I missed something in the long thread, but I didn't saw any actual critique on the app itself. No bugs, no code quality problems pointed out. All critique seems to stem from people, who didn't even look at it, based on their prejudices and feelings for tools they never looked it instead of based on the actual thing. Almost like a LLM hallucinating something together because it is missing information.

#280 makes it even more ridiculous. It feels like in the lack of actual problems with the thing, they are now just making up arbitrary rules aimed at exactly this thing. 
> As a community largely composed of passionate developers and reverse engineers, we

Who is "we" for whom are you speaking? For Maschell? For rw? For Crementif? For me? For any of the many other developers of homebrew for the Wii U you are claiming to speak for? Or are you just projecting your feelings on them and assume that's what they think?

Whom does this rule actually serve? Does it serve the developers? Is it in the interest of the users? Does it solve a problem that actually happened? Or is it just to support your feelings?

As far as I can tell no one here ever cared for the maintainability of projects used here. And there is lots of bad human written code out there. Why pretend you care now? Also even if if someone can maintain their code, who says they will? Remember every software here is provided as is, with no guarantees and no obligations on the developer. 

Once this reached stable and does what it promised and I went trough the code I will likely include it on my ISFShax guide for cases where the Browser still works. The website is by someone I trust anyway and who confirmed the ownership of the site to me. So I don't have concerns that there will be something evil in the binary.

**Comment by zer00p** [comment]
This was brought to my attention: https://heyquark.com/brew/2026/02/27/on-ai-homebrew/ and I have a few thing to say about this.
First of all thanks to the author @ashquarky for actually taking the time and reviewing the code.

## The fw_img_loader attribution / LICENSE

There I messed up and there is no excuse for that and I can't even blame a LLM for that. So read this as explanation what happened, not as an attempt to excuse that. It was a failure of me period. 
It happened already in the second commit I did: https://github.com/zer00p/wafel_installer/commit/53e048eb218f47d531f50a0f68752bb21a8572c1 . I took the patches I was interested in indeed from the fw_img_payload. I wasn't interested in the repair data so I skipped them, and with that the notice above it.

I corrected this already yesterday when I was also adding attributions for other Homebrew projects I drew inspiration and code from / gave gemini as example (Aroma Updater). In this process I also converted the Project to GPLv2, which is needed to comply with the GPLv2 of the fw_img_loader and other parts of Aroma.

## The patches

Discussing this is a good opportunity to give some insights on how I actually use these tools. As already said I copied the relevant patches from the fw_img_payload and provided them either as a base file or in the prompt to gemini cli. Of course I had to explain gemini what kind of addresses to use, what the address calculation there means and what mechanism to use instead, since we can't just write the IOSU memory from the PPC. Sadly I only committed the working result, so part of the transformation is lost.  But in that first working code you can still see the original comments and that it seems very verbose and paranoid about the KernWrite failing, which should never happen. So why is this? Simple answer: I had to debug, it was just freezing before even doing a reload. - Which is strange since other KernWrites worked and the patches themselves where known good too.

So what to do? First remove what isn't strictly needed and could cause problems. There is a set of 3 patches going to IOSU Crypto. We know that the crypto module changed between 5.5.0 and 5.5.5. My IOSU decomp is based on 5.5.0, so I couldn't just look if they still line up. This was only intended to load the isfshax installer, which is not encrypted and the code path to skip the decryption also skips the hash checks that were "noped" or should I more accurately say "moved" out.
This didn't solve the problem, but I didn't see a reason to put them back, since not including patches, that you don't need and you can't verify what they are overwriting is safer than to keep them.
- Side note: Due to "feature creep" I later added an option to load arbitary fw.img from SD, if someone didn't have stroopwafel installed and would try to load a custom encrypted fw.img, it would likely fail due to the hash checks. It's noting dangerous, but you would wonder why it didn't work and annoy that you have to get up and force the console off. So I might consider adding them back.

But we still have the freezing problem... So what to do now? Just add more logging of course. That's where the prints originated with the `launch_os_hook bl`, which quarky took issue with (and ofc I had gemini add them for me, because why would I do something that trivial and tedious by hand?):
> The human author may have had access, but even so they didn’t catch this in review, or didn’t care

And yes I can tell you I couldn't care less what these strings exactly said, as long as I could tell them apart. Also I could argue that `launch_os_hook bl` makes grammatically sense if you read it as a compound word. It's the `bl` for the `launch_os_hook`, but truth is I just didn't care about a string which sole purpose was to help me debugging.

So where did it crash? Here: 
```C
Mocha_IOSUKernelWrite32(0x050282AE, 0xF031FB43); // bl launch_os_hook
```
Probably you  can already guess what the problem is. If not, I can hint you that the next commit is related to it: https://github.com/zer00p/wafel_installer/commit/73d0f463680a93d055985346da306eedc61bd8ad 
The commit message should give it away now... But in case you are not familiar with ARM: on ARM word (4byte/ 32bit) memory accesses have to be 4 byte aligned when the MMU is enabled (which it normally is). If we look at the memory address, we see it is only 2 byte aligned: `0x050282AE % 4 = 2`. (slight tangent) Now you might be wondering: "We are patching a 4 byte instruction here, arm uses 4 byte instructions, surely they have to be aligned." and you would be almost right. In the normal ARM mode that would be the case, but there is a so called thumb mode, which uses 2 byte instruction. The bl is a special case, in thumb mode it consists in reality of 2 2byte instructions).
So why did it work on the fw_img_payload? It just disables the MMU:
```C
unsigned int control_register = disable_mmu();
```
My first fix was then to patch the dumpling internal "CFW", which provides the KernWrite. Simply using a memcpy, which handles the unaligned word write, by splitting it in multiple smaller writes as needed.

And that brings us to:
> The same patches are made in the same order, with the exception of the first

I assume this is about this very patch we have been talking about
So why was it changed after this first version which I got working and still the original 4 byte patch? I might have patched the internal CFW, but we can't rely on having that. When launched from Aroma, it will use the Mocha that ships with aroma. And this Mocha doesn't have my patch for the unaligned access.
So how do we support that?
We simply do two 4 byte patches which are then 4 byte aligned to cover the 4 byte region we are interested in. But now we are overwriting more than the original path. So what are we writing there? Something that the AI hallucinated? Let me just have a look at Ghidra:
<img width="390" height="54" alt="Image" src="https://github.com/user-attachments/assets/9760a183-a3f1-4be6-a7fc-10fa0dd0dcfa" />
And compare that to the modified patch:
```C
uint64_t p2 = 0x00a4F031FB43193b; // need to align patch for old mocha
if (!applyPatch(0x050282AC, &p2, 8, L"Applying patch 2 (launch_os_hook bl)...")) return;
```
Did AI by accident hallucinate exactly the same values as the original ones? Or did gemini look at the binary?
Also why does the casing of the first and the last halfword fit the casing in Ghidra, while the middle word has the casing of the original patch? Arent LLMs known to be very consistent about that?

Then we also had the "nop nop" incident:
> Ancast header nop nop is kinda just meaningless

The original had this:
```C
// remove various Ancast header size checks (somehow needed for unencrypted fw.img)
*(u32*)(0x0500A7C8 - 0x05000000 + 0x081C0000) = 0xbf00bf00; // nop nop
*(u16*)(0x0500A7C8 - 0x05000000 + 0x081C0000) = 0xe00f; // b #0x500a7ea
```
As you can see both go against the same memory address, so kind of redundant. If we looks closely the second one is only a 2 byte write, overwriting half of the first one. The 2byte write is problematic, since the KernWrite function is expecting a full 4 byte word. Gemini cooked something up where it first read the original value back (the one it just wrote) masking and combining it before writing it back. Any sane human would have just combined them:
```C
    WHBLogFreetypePrint(L"Applying patch 9 (Ancast header nop nop)...");
    status = Mocha_IOSUKernelWrite32(0x0500A7C8, 0xe00fbf00); // b #0x500a7ea
```
The nop nop in the string was just a residue from the original, which really were two nops. Have you ever seen an outdated comment, that didn't match the code anymore?

Then we have
> The new, bespoke applyPatch system

which was only introduced later. Here: https://github.com/zer00p/wafel_installer/commit/edcd0297cae6e122ec08fe57f7d84587d6bf46ab

So why did Jules do that? Because I told it to. I don't like code duplication. I understand the arguments why you might duplicate some code, but this is my project and that's how I like it.

Also it seems quarky missed one patch:
```C
MochaUtilsStatus status = Mocha_IOSUKernelRead32(0x0501f578, &val);
if (status != MOCHA_RESULT_SUCCESS) {
    WHBLogFreetypePrint(L"Warning: Failed to read ISFShax fallback patch!");
    WHBLogFreetypeDrawScreen();
} else if (val == 0x32044bcc) {
    uint32_t undo_val = 0xe0914bcc;
    if (!applyPatch(0x0501f578, &undo_val, 4, L"Undoing ISFShax fallback reload patch...")) return;
}
```
This is needed in case ISFShax is installed, but stroopwafel isn't. In that case ISFShax applies a set of fallback patches that make the console boot. But one of these patches causes the console to do a full reboot instead of a reload. That makes perfect sense under normal circumstances, since if it reloaded the IOSU, it would be missing the ISFShax patch and crash, but here we need the reload to load the custom fw.img.

> I hope I have proven that the system which wrote this code did not actually read the IOSU assembly they are patching and modifying.

How does it then come that all the modifications serve a certain purpose and the modifications even have comments that explain them? "// need to align patch for old mocha"

> Do you feel an AI system, which only references existing code and does no original research, is trustworthy for this?

Yes, as the system did exactly what it was told to.

> This tool doesn’t have access to the binary this code modifies, and doesn’t seem to have researched the earlier history or lineage. I will show it embarked on a refactor even though the tool doesn’t understand what the code is doing.

If you make such a claim, you should make sure you are correct on such a statement (looking at the date of the patches quoted here):

> Now, here is a snippet from Aroma’s fw_img_loader payload. This code dates from early 2020.

<img width="1142" height="655" alt="Image" src="https://github.com/user-attachments/assets/ae901a4e-6e98-4a09-ad34-e8b37d1375bf" />

And these were the reason why this was chosen as the base: it added the capability to load the unencrypted isfshax installer.


## Safty

Keeping this short, since it post is already long enough...

> The vibe-coded project [downloads isfshax from GitHub](https://github.com/zer00p/wafel_installer/blob/master/source/app/download.cpp#L260) and executes it without checks.

Which is also what the current guide does. Since the checks are implemented in the isfshax installer (which gets used here too)

> What happens if the repository URL changes?

Then the download will just fail, the app will show an error end. But ISFShax been stable for years. I am confident it will stay where it is. If that remote case ever happens, then the app would get an update. Adding unnecessary complexity, without a clear need for such a remote case would add more risk imo. Especially since nothing catastrophic would happen. 

> What happens if the CA cert is bad?

What exactly should that mean? It is embedded in the app, together with a up to date curl to support the latest TLS standard. How would the CA cert suddenly be bad? But if it were all downloads would just fail and nothing would happen.

> What happens if the user’s WiFi drops out?

then the download fails and the user has to press retry or can abort if they want.

> SD card is corrupt?

Then he needs a new SD / reformat it. Same as with the current guide. The superblock image gets stored on the SLC not the SD. If the superblock.img got corrupt, then it wouldn't match the superblock.sha file and the installer would reject that. I even verified that.

> Hynix MLC dies?

Then I hope he used the wafel installer instead of the hacks.guide to setup their homebrew so they ISFShax.

> We keep getting new versions of isfshax but isfshax_installer moves to a different repository?

The isfshax authors are aware of this and this won't happen. If the superblock gets changed so that installing it with an outdated installer would be catastrophic, then they would have to change the hashing anyway to also protect against other cases where users might miscombine them.

> What if we get different versions for different OS releases? 5.5.4 vs. 5.5.5?

ISFShax only cares for the boot1. The ISFShax installer checks the boot1 version If the firmware is so old that the boot1 is different, then also the browser and IOSU exploit won't work.

> The lack of hash checks or pinning is probably fine

The hash is checked by the installer. If any other file corrupts, it won't have catastrophic consequences.
Also care was taken to download the fw.img only after the rest of stroopwafel is there. If it wasn't, it wouldn't be catastrophic, but we still want to avoid that.

> no matter that Aroma Updater already sets a standard for hash and version-checking every file.

I appreciate what Maschell is doing there. I agree that you should avoid unnecessary risk but at some point you have to say it is good enough, since it will never be perfect. You are always taking a risk. Even if you do nothing or if you hold something back out of fear, which could have safed someone from some loss. What is if mochapayload corrupts on the SD and writes some patch to the ISFS FAT buffer instead of the correct location? Same for the pretendo plugin when it patches IOSU. Should we establish hash checking on every binary that we load before executing it?
Most problems I have seen with Aroma updates were due to corrupted FAT32 file systems. One cause for that was homebrew recommended by the guide, which double mounts the SD and no one cared for years.
Also how is it responsible that the guide is telling people to download some piece of software, from a sketchy website, that doesn't even have https and run it with admin priviledges. That can cause real harm to people, since other than a old game console, a PC can hold sensitive information or information of much value, like tax data. Some people rely on their PC for their business. So how is that responsible? How is it responsible to start server on every console, that listens on the network and just waits to execute commands without any authentication?

So if you think you found something inherently dangerous in this software, please tell me. I think I did my due diligence to make sure it is not inherently dangerous. But of course you can also miss something or don't think of some case. But that is independent of if you use AI or not.
The critical stuff is encapsulated in the isfshax installer, which brings it's own complete environment. We are not overwriting or touching system files. Downloads use https. ISFShax has fallback modes.

Currently I was working on making the downloads more reliable by downloading to a temp location first and making sure files don't exceed a certain size when written to SLC, although I don't deem it strictly necessary here. There also already exists code to fetch the shas from github, which is used to check the installed files for updates, but I am not planning on using that for verifying downloads. For the criticality of the files here the protection offered by https is enough imo. If someone really wants it, it can be considered - but then I would also like the guide to tell people to check the hashes of the files they download from the aroma website.


## Energy Consumption

in the month I am now working on this I used about 1100 minutes of Jules. Assuming something around 100K tokens per minute (based on statistics from geminicli) of gemini pro on TPUv7, this burned around 50kWh (ofc that might vary) on my free time project in a month. From that I could drive around 300km with a efficient BEV. I would say that is significant. But would you tell someone they aren't allowed to use their car for their free time activities? I didn't use a car in the last month...
And this was certainly one of the more extreme months. 
Looking at CO2: if this prevents two people from from buying an extra HDD, because they can use their existing SD card, then the CO2 produced by this is already offset.

**Comment by team-orangeBlue** [comment]
Babble bobble, om nom nom, hurr durr, pfft bzzt.

So this thread has turned into "ai sux 🍅" from "when will we get ISFShax on the guide".

Let's ignore that automated installer and ask again: when will we get ISFShax on the guide?

It's literally a reference to GBAtemp, or a mention of https://isfsh.ax, copypaste and forcing everyone to get fwimg loader, etc, literal cake walk, I might as well do it ~~next week or something~~ at some point in time - what's the deal?

**Comment by Lazr1026** [comment]
I don't know if jan plans to do any more updates to the isfshax superblocks, and I don't exactly know what has been done (on the Aroma side) to seamlessly integrate Aroma into the isfshax ecosystem.

**Comment by jan-hofmeier** [comment]
There is still that one update that is pre release. Although I tested it on wup 01, I don't have Wup 50s to test. 
~~Also I want to change the behaviour of holding the power button from forcing SD load to forcing fallback mode. 
Since SD boot is the first thing it tries anyway, there is no real need to keep the function to force that anymore. And having a way to force fallback mode could become helpful in the context of the wafel installer. Imagine someone messing up the SLC plugins and having fastboot minute and no SD because they are using the SD Emulation. They would have to buy them an SD another wise.~~ (just noticed, I already implemented that, as soon as somone tests the superblock with a wup 50, I can release it)
Fastboot will also get another update to go to fallback mode if loading plugins fails for any reason.
The last update to the ISFShax installer were specifically to support an automated install

**Comment by zer00p** [comment]
https://wafel.xyz should now be pretty feature complete. Even with every feature tested I wouldn't be surprised if there is still some bug hiding in one of the many possible paths a user can take. So would appreciate bug reports.
It should be obvious, but here I say it again: it is not doing anything unsafe.
